Feature/list dictionary

BI003 opened the pull request.

BI003 committed:
Реализовать сервис «TestDataGenerator»,
предоставляющий методы:
"GenerateClient()" - генерации коллекции 1000 клиентов банка;
"GenerateClientsDictionary()" - генерации словаря в качестве ключа которого применяется номер
телефона клиента, в качестве значения сам клиент;
"GenerateEmployee()" - генерации коллекции 1000 сотрудников банка.

Так же для заполнения коллекций были созданы генераторы имён, фамилий, номеров телефонов и номеров паспортов.

BI003 committed:
В папке “Services” реализован сервис «TestDataGenerator», предоставляющий методы:
"GenerateClients" - генерации коллекции 1000 клиентов банка;
"GenerateClientsDictionary" - генерации словаря в качестве ключа которого применяется номер телефона клиента, в качестве значения сам клиент;
"GenerateEmployees" - генерации коллекции 1000 сотрудников банка.

Осуществил:
1) Пользуясь инструментом “Stopwatch”, провел замер времени выполнения поиска клиента по его номеру телефона среди элементов списка;
2) Провел замер времени выполнения поиска клиента по его номеру телефона, среди элементов словаря;
3) Выборку клиентов, возраст которых ниже определенного значения;
4) Поиск сотрудника с минимальной заработной платой;
5) Сравнил скорость поиска по словарю двумя методами:
А) При помощи метода LastOrDefault(ищем последний элемент коллекции);
Б) Найти в словаре этот же элемент, по ключу.

BI003 committed:
Убрал ненужные using

Dem-Ivan commented on BankSystem.App/Services/TestDataGenerator.cs:
```diff
@@ -0,0 +1,58 @@
+            var faker = new Faker<Client>("ru")
+                .RuleFor(c => c.Name, f => f.Name.FirstName())
+                .RuleFor(c => c.Surname, f => f.Name.LastName())
+                .RuleFor(c => c.Number, f => f.Random.Int(3731000, 3732000))
```
Плохое название Number, вредил в нем хранится номер телефона, в лучшем случае порядковый номер. Вот ели бы он назывался Phone или PhoneNumber, тогда другое дело.

У богуса есть собственные инструменты для генерации номеров, адресов, и прочее...

Dem-Ivan commented on BankSystem.App/Services/TestDataGenerator.cs:
```diff
@@ -0,0 +1,58 @@
+                var client = faker.Generate();
+                if (!clientDictionary.ContainsKey(client.Number))
+                {
+                    clientDictionary[client.Number] = client;
```
Список можно преобразовать в словарь одним выражением. Изучите этот вопрос.

Dem-Ivan commented on Practice/Program.cs:
```diff
@@ -52,5 +53,72 @@ static void Update(ref Currency currency, int price)
+        else
+        {
+            Console.WriteLine("Словарь клиентов пуст.");
+        }
```
1) Вся эта обвязка с проверками и выводом в консоль, исказит результаты замеров! Внутри замера ее быть не должно! Кроме того, замер как правило проводят статистически, многократно, в цикле. Первый результаты считают ложными, списывая их на "прогрев" системы.

2) Не вижу сравнения с поиском по ключу.

Dem-Ivan commented on BankSystem.App/Services/TestDataGenerator.cs:
```diff
@@ -0,0 +1,58 @@
+        public List<Client> FindYoungerClients(List<Client> clients, int limitAge)
+        {
+            return clients.FindAll(client => client.Age < limitAge);
+        }
```
Методы слиплись как пельмени после варки

Dem-Ivan commented on BankSystem.App/Services/TestDataGenerator.cs:
```diff
@@ -0,0 +1,58 @@
+        }
+        public List<Client> FindYoungerClients(List<Client> clients, int limitAge)
+        {
+            return clients.FindAll(client => client.Age < limitAge);
```
Такого метода в сервисе генерации быть не может. Иначе это не генератор а Швейцарский нож.

BI003 committed:
Выполнил работу над замечаниями:
Изменил название Number на PhoneNumber.
Использовал инструмен Bogus для генерации номера телефона.
Преобразовал список в словарь одним выражением.
Перемести метод "FindYoungerClients" в Bank.Service.
Визуально отделил методы друг от друга в "TestDataGenerator".
Удалил все проверки и выводы в консоль, искажающие результаты замеров.
Провёл все замеры в цикле.
